src/prefect/engine/result_handlers/s3_result_handler.py

```diff
@@ -63,8 +65,9 @@ def initialize_client(self) -> None:
 
     @property
     def client(self) -> "boto3.client":
-        if not hasattr(self, "_client"):
+        if not prefect.context.get("boto3client"):
             self.initialize_client()
+            prefect.context["boto3client"] = self._client
```

Comment:
Could you elaborate on why you are putting this in context instead of referencing the attribute?  I can see why this works, but in general modifying context outside of a runner pipeline could lead to unexpected behavior or in this case redundant computation (e.g., if this gets added to context deep in a `TaskRunner` pipeline, it won't be available to the next task because `context` is generally read-only).

Reply:
I thought the point was to reinstantiate the session AND client anytime we were in a new thread, which could be per task run (and thus in my mind, per Prefect context). Since you’re asking I’m assuming I’m misunderstanding what “global boto session” means in the issue OP and/or how exactly resources are shared between tasks on dask

Reply:
Yea, that definitely makes sense; my knee-jerk reaction is that modifying context outside of a pipeline is a smell, but in this case because the context is thread-local I agree it's the cleanest way to detect when running in a new thread so I'm 👍

Reply:
@cicdw Gotcha yeah I had similar misgivings but couldn't figure out another way from the perspective of result handler to determine if it was likely we were in another thread without adding a dependency (and myself possibly needing to know more) about executor submission.

I will update the changelog, but re: testing do you mind commenting on whether
> Should I be attempting to add an integration test testing this edge between Dask executor, mapped flows and S3?

Reply:
Hmmm at this moment I'd say "no" - the overhead of mocking out the s3 calls across the threads / processes feels too heavy to me, unless you see a clear way to do it.

Reply:
Gotcha thanks -- I don't see a clear way commensurate with the PR size, but I can add it to some discussion points for us tomorrow at blue sky. Thanks 🙏